File: backend/go-fiber/internal/middlware/authentication.go
package middleware

import (
	"MarkroCounter/internal/dao"
	"MarkroCounter/internal/helper"
	"MarkroCounter/internal/types"
	"context"
	"encoding/json"
	"errors"
	"github.com/gofiber/fiber/v2"
	"go.mongodb.org/mongo-driver/mongo"
	"net/http"
	"strings"
)

func OIDCAuthMiddleware() fiber.Handler {
	userinfoEndpoint := "https://" + helper.Ev("AUTH_DOMAIN") + "/api/oidc/userinfo"
	return func(c *fiber.Ctx) error {
		authHeader := c.Get("Authorization")
		if authHeader == "" {
			return helper.SendUnauthorized(c, errors.New("missing Authorization header"))
		}

		// Extract the token from the Authorization header
		token := strings.TrimPrefix(authHeader, "Bearer ")
		if token == authHeader { // No "Bearer " prefix found
			return helper.SendUnauthorized(c, errors.New("invalid Authorization header"))
		}

		// Fetch userinfo from Authelia userinfo endpoint
		user, err := fetchUserInfo(userinfoEndpoint, token)
		if err != nil {
			return helper.SendUnauthorized(c, err)
		}

		err = storeUserInDB(c, user)
		if err != nil {
			return err
		}

		// Store relevant user information in Locals
		c.Locals("user", user)

		// Proceed to the next handler
		return c.Next()
	}
}

func storeUserInDB(c *fiber.Ctx, user *types.User) error {
	db := c.Locals("db").(*mongo.Database)

	_, err := dao.UpsertUser(db, user)
	if err != nil {
		return err
	}

	role, err := setRoleToUserIfNotSet(err, db, user)
	if err != nil {
		return err
	}
	user.Role = role

	return nil
}

func setRoleToUserIfNotSet(err error, db *mongo.Database, user *types.User) (types.Role, error) {
	role, err := dao.GetUserRole(db, user.PreferredUsername)
	if err != nil {
		return types.NoRole, err
	}

	if role == types.NoRole {
		_, err = dao.SetUserRole(db, user.PreferredUsername, types.RoleUser)
		if err != nil {
			return types.NoRole, err
		}
	}
	role = types.RoleUser
	return role, nil
}

// fetchUserInfo fetches the userinfo from Authelia using the provided token
func fetchUserInfo(userinfoEndpoint, token string) (*types.User, error) {
	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, userinfoEndpoint, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+token)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, errors.New("failed to fetch userinfo: " + resp.Status)
	}

	user := &types.User{}
	if err := json.NewDecoder(resp.Body).Decode(user); err != nil {
		return nil, err
	}

	return user, nil
}

File: backend/express/src/middleware/Authentication.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OIDCAuthMiddleware = void 0;
const axios_1 = __importDefault(require("axios"));
const User_1 = require("../types/User");
const UserDao_1 = require("../dao/UserDao"); // Definiere deine User- und Role-Typen hier
const OIDCAuthMiddleware = (req, res, next) => {
    const userinfoEndpoint = `https://${process.env.AUTH_DOMAIN}/api/oidc/userinfo`;
    const authHeader = req.headers.authorization;
    if (!authHeader) {
        res.status(401).send('Authorization header missing');
        return;
    }
    const token = authHeader.replace('Bearer ', '');
    if (token === authHeader) {
        res.status(401).send('Invalid Authorization header');
        return;
    }
    fetchUserInfo(userinfoEndpoint, token)
        .then(user => {
        if (!user) {
            res.status(401).send('Invalid token');
            return;
        }
        const db = res.locals.db;
        storeUserInDB(db, user).then(() => {
            res.locals.user = user;
            next();
        });
    })
        .catch(err => {
        next(err);
    });
};
exports.OIDCAuthMiddleware = OIDCAuthMiddleware;
const storeUserInDB = (db, user) => __awaiter(void 0, void 0, void 0, function* () {
    yield (0, UserDao_1.upsertUser)(db, user);
    let role = yield (0, UserDao_1.getUserRole)(db, user.preferredUsername);
    if (role === User_1.Role.NOROLE) {
        yield (0, UserDao_1.setUserRole)(db, user.preferredUsername, User_1.Role.USER);
        role = User_1.Role.USER;
    }
    user.role = role;
});
const fetchUserInfo = (userinfoEndpoint, token) => __awaiter(void 0, void 0, void 0, function* () {
    const response = yield axios_1.default.get(userinfoEndpoint, {
        headers: {
            Authorization: `Bearer ${token}`
        }
    });
    if (response.status !== 200) {
        console.error(`Failed to fetch userinfo: ${response.statusText}`);
        return null;
    }
    return response.data;
});
